packages/cells/src/model.ts

```diff
@@ -790,8 +825,23 @@ export class CodeCellModel extends CellModel implements ICodeCellModel {
   ): void {
     if (change.outputsChange) {
       globalModelDBMutex(() => {
-        this.outputs.clear();
-        slot.getOutputs().forEach(output => this._outputs.add(output));
+        let retain = 0;
+        for (const outputsChange of change.outputsChange!) {
+          if ('retain' in outputsChange) {
+            retain += outputsChange.retain!;
+          }
+          if ('delete' in outputsChange) {
+            for (let i = 0; i < outputsChange.delete!; i++) {
+              this._outputs.remove(retain);
+            }
+          }
+          if ('insert' in outputsChange) {
+            // Inserting an output always results in appending it.
+            for (const output of outputsChange.insert!) {
+              this._outputs.add(output.toJSON());
```

Comment:
This throws an error when used with `jupyter-server-nbmodel`:

```
TypeError: e.toJSON is not a function
    at jlab_core.abd83dd7cfacd6da4d00.js?v=abd83dd7cfacd6da4d00:1:238328
    at 1168.8e5df3139aa020a611cb.js?v=8e5df3139aa020a611cb:1:551
    at q._onSharedModelChanged (jlab_core.abd83dd7cfacd6da4d00.js?v=abd83dd7cfacd6da4d00:1:238119)
    at m (jlab_core.abd83dd7cfacd6da4d00.js?v=abd83dd7cfacd6da4d00:1:1838868)
    at Object.l [as emit] (jlab_core.abd83dd7cfacd6da4d00.js?v=abd83dd7cfacd6da4d00:1:1838544)
    at a.emit (jlab_core.abd83dd7cfacd6da4d00.js?v=abd83dd7cfacd6da4d00:1:1836381)
    at Array._modelObserver (1168.8e5df3139aa020a611cb.js?v=8e5df3139aa020a611cb:1:7524)
    at Object.r (383.db345dbeef5ef774e50c.js?v=db345dbeef5ef774e50c:1:936)
    at ji (383.db345dbeef5ef774e50c.js?v=db345dbeef5ef774e50c:1:43499)
    at Array.<anonymous> (383.db345dbeef5ef774e50c.js?v=db345dbeef5ef774e50c:1:52559)
```

It was working with JupyterLab 4.2. I am not sure which package is breaking the API contract yet.

Reply:
This is still happening on `main` after https://github.com/jupyterlab/jupyterlab/pull/16692 was merged

Reply:
I see the latest change there was one month ago, and https://github.com/jupyterlab/jupyterlab/pull/16498 was merged two weeks ago, so isn't that expected?

Reply:
I think it is not expected in the sense that this looks like a breaking API requirement change. This is previously it accepted JSON and now it requires an object with `toJSON()` method? I am trying to understand this really. I tested locally that a simple change of `'toJSON' in output ? output.toJSON() : output` makes it work.

Reply:
Previously we had:
```js
    if (change.outputsChange) {
      globalModelDBMutex(() => {
        this.outputs.clear();
        slot.getOutputs().forEach(output => this._outputs.add(output));
      });
    }
```
So any change to the outputs resulted in clearing them and adding the updated outputs again, which was very inefficient.
Now we have:
```js
    if (change.outputsChange) {
      globalModelDBMutex(() => {
        let retain = 0;
        for (const outputsChange of change.outputsChange!) {
          if ('retain' in outputsChange) {
            retain += outputsChange.retain!;
          }
          if ('delete' in outputsChange) {
            for (let i = 0; i < outputsChange.delete!; i++) {
              this._outputs.remove(retain);
            }
          }
          if ('insert' in outputsChange) {
            // Inserting an output always results in appending it.
            for (const output of outputsChange.insert!) {
              this._outputs.add(output.toJSON());
            }
          }
        }
      });
    }
```
So we really interpret the changes and do things in a finer grained manner.
`this._outputs.add` accepts a JSON, just that before we were converting all the shared outputs to JSON in `slot.getOutputs()` and now we do it only for the inserted outputs.